spring/spring4/spring4-common/src/main/java/org/infinispan/spring/provider/CacheDelegate.java

```diff
@@ -71,13 +75,38 @@ public ValueWrapper get(final Object key) {
 
    @Override
    public <T> T get(Object key, Callable<T> valueLoader) {
-      return (T) nativeCache.computeIfAbsent(key, keyToBeInserted -> {
-         try {
-            return valueLoader.call();
-         } catch (Exception e) {
-            throw ValueRetrievalExceptionResolver.throwValueRetrievalException(key, valueLoader, e);
+      long stamp = synchronousGetLock.readLock();
+      try {
+         T value = (T) nativeCache.get(key);
+         if (value == null) {
+            while (true) {
+               long upgradedLockStamp = synchronousGetLock.tryConvertToWriteLock(stamp);
```

Comment:
Is the loop really necessary? The way I see it, either the upgrade succeeds the first time and you get a write lock, or it didn't succeed and `writeLock()` blocks until you get a write lock.

Reply:
This works slightly different than that. Here are relevant parts from [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/StampedLock.html):

> ...The state of a StampedLock consists of a version and mode. Lock acquisition methods return a stamp that represents and controls access with respect to a lock state; "try" versions of these methods may instead return the special value zero to represent failure to acquire access...

> ...This class also supports methods that conditionally provide conversions across the three modes. For example, method tryConvertToWriteLock(long) attempts to "upgrade" a mode, returning a valid write stamp if (1) already in writing mode (2) in reading mode and there are no other readers or (3) in optimistic mode and the lock is available...

So effectively if you want to achieve upgrade behavior, you need to use `tryConvertToWriteLock` and pass read lock stamp into that method. With `writeLock` you just get another lock for writing.

BTW, if you look into `else` block, which happens when I can't upgrade the lock, I just obtain a write lock and do stuff in the next iteration.